Enable test reruns on failed fragiled tests

avtikhon opened the pull request:
Added ability to use fragile list in JSON format from 'suite.ini'
files.
```
   fragile = {
        "retries": 10,
        "tests": {
            "bitset.test.lua": {
                "issues": [ "gh-4095" ],
                "checksums": [ "050af3a99561a724013995668a4bc71c", "f34be60193cfe9221d3fe50df657e9d3" ]
            }
        }}
```

Added ability to set per suite in suite.ini configuration file
'retries' option, which sets the number of accepted reruns of
the tests failed from 'fragile' list.

Added ability to check failed tests w/ fragile list to be sure that
the current fail equal to the issue mentioned in the fragile list.

Closes #189.

LeonidVas commented on lib/worker.py:
```diff
@@ -313,7 +315,7 @@ def run_task(self, task_id):
         return short_status
 
-    def run_loop(self, task_queue, result_queue):
+    def run_loop(self, task_queue, result_queue, fragile_retries):
```
fragile_retries is saved in self.suite.ini, may be forward all fragile stuff through arguments is not a good idea, may be just pass the is_fragile flag?

avtikhon replied:
Right, removed fragile_retries and used self.suite.fragile_retries() where was needed. Set is_fragile here.

LeonidVas commented on lib/worker.py:
```diff
@@ -326,7 +328,19 @@ def run_loop(self, task_queue, result_queue):
+            short_status = ''
+            # 'fragile_retries' value comes from the suite configuration
+            retries_left = fragile_retries
+            while short_status != 'pass' and retries_left >= 0:
```
Must be used only for "fragile" tests. IMHO, retries_left > 0 looks more correct.
May be something like:
run_test()
if no pass and fragile then retry...

avtikhon replied:
Split into 2 parts for fragile and not fragile.

LeonidVas commented on lib/worker.py:
```diff
@@ -340,14 +354,14 @@ def run_loop(self, task_queue, result_queue):
             self.task_done(task_queue)
 
-    def run_all(self, task_queue, result_queue):
+    def run_all(self, task_queue, result_queue, fragile_retries):
```
The same as for run_loop() method.

avtikhon replied:
Right, corrected.

LeonidVas commented on lib/test_suite.py:
```diff
@@ -174,6 +174,14 @@ def fragile_tests(self):
                 res.append(test)
         return res
 
+    def fragile_checksums(self):
```
Maybe fragile_checksums()  should return a dict {test_name_1 -> [checksum1, checksum2],
test_name_2 -> [checksum1]}

avtikhon replied:
Sure, added ability to use JSON format.

LeonidVas commented on lib/worker.py:
```diff
@@ -83,6 +84,7 @@ def get_task_groups():
                 'task_ids': stable_task_ids,
                 'is_parallel': suite.is_parallel(),
                 'fragile_retries': 0,
+                'fragile_checksums': suite.fragile_checksums(),
```
As I proposed, just add the "is_fragile" flag.

avtikhon replied:
Right, added flag is_fragile.

LeonidVas commented:
It looks like this closes # 189. Add this information to the commit message instead of "Part of tarantool / tarantool problem # 5050".

avtikhon commented on dispatcher.py:
```diff
@@ -355,6 +355,7 @@ def __init__(self, key, task_group, randomize):
         self.gen_worker = task_group['gen_worker']
         self.task_ids = task_group['task_ids']
         self.is_parallel = task_group['is_parallel']
+        self.fragile_retries = int(task_group['fragile_retries'])
```
Right, check is not needed here.

avtikhon commented on lib/test_suite.py:
```diff
@@ -117,6 +117,10 @@ def __init__(self, suite_path, args):
         self.parse_bool_opt('is_parallel', False)
         self.parse_bool_opt('show_reproduce_content', True)
 
+        fragile_retries = self.ini.get('fragile_retries')
```
Created self.parse_int_opt() method.

avtikhon commented on lib/worker.py:
```diff
@@ -315,7 +318,7 @@ def run_task(self, task_id):
         return short_status
 
-    def run_loop(self, task_queue, result_queue, fragile_retries):
+    def run_loop(self, task_queue, result_queue, fragile_retries, fragile_checksums):
```
Right, removed.

avtikhon commented:
> It looks like this closes # 189. Add this information to the commit message instead of "Part of tarantool / tarantool problem # 5050".

Ok, changed.

LeonidVas commented:
After some discussion with @Totktonada and @avtikhon, it was decided to try using JSON for configuration a "fragile".
Example:
```
fragile =
 {
   "tests": {
     "test_name": {
       "checksums": [
         ch1,
         ch2
       ]
     }
   },
   "retries": 10
 }
```
Useful links:
https://docs.python.org/3/library/configparser.html#supported-ini-file-structure
https://tools.ietf.org/html/rfc7159

avtikhon commented:
> After some discussion with @Totktonada and @avtikhon, it was decided to try using JSON for configuration a "fragile".
> Example:
> 
> ```
> fragile =
>  {
>    "tests": {
>      "test_name": {
>        "checksums": [
>          ch1,
>          ch2
>        ]
>      }
>    },
>    "retries": 10
>  }
> ```
> 
> Useful links:
> https://docs.python.org/3/library/configparser.html#supported-ini-file-structure
> https://tools.ietf.org/html/rfc7159

Implemented.

Totktonada commented:
I didn't look thoroughly, but I would highlight several points.

1. The approach with checksums is not universal. It does not work if time-related information appears in a result file. It requires periodical checksums updating if a line number is reported as part of an error. I guess it is okay to operate on a test level and either mark it as stable or as fragile.
2. We can either implement reruns within a worker or on the level upward: within the task management system (dispathers-listeners-workers). Since there is nothing tarantool-specific I would look into the latter way. I pushed a variant to `Totktonada/rerun-fragile-test` branch (0292fe2ad901f33c903387f5498ade4f13acef94).

----

If you want my opinion about extracting checksums from comments — it is disgusting. Sorry.

There are more points I would highlight about the patchset, but it does not have much sense if we'll get rid of checksums at all. The patchset modifies many different parts of test-run: too much for this small feature, IMHO.

avtikhon commented on lib/test_suite.py:
```diff
@@ -110,13 +122,23 @@ def __init__(self, suite_path, args):
         self.parse_bool_opt('is_parallel', False)
         self.parse_bool_opt('show_reproduce_content', True)
 
+        self.parse_int_opt('fragile_retries', 0)
```
Removed.

avtikhon commented on lib/test_suite.py:
```diff
@@ -154,22 +176,44 @@ def collect_tests(self):
         return self.tests
 
+    def get_fragile_list(self):
+        if self.ini['fragile_is_json']:
```
Removed.

avtikhon replied:
Removed.

avtikhon commented on lib/test_suite.py:
```diff
@@ -154,22 +176,44 @@ def collect_tests(self):
 
+    def get_fragile_checksums(self):
+        checksums = []
+        if self.ini['fragile_is_json']:
```
Removed.

avtikhon commented on lib/test_suite.py:
```diff
@@ -250,5 +294,14 @@ def run_test(self, test, server, inspector):
         return self.ini['is_parallel']
 
+    def fragile_retries(self):
+        if self.ini['fragile_is_json']:
```
Removed.

avtikhon commented on lib/worker.py:
```diff
@@ -326,9 +330,30 @@ def run_loop(self, task_queue, result_queue):
 
             result_queue.put(self.current_task(task_id))
-            short_status = self.run_task(task_id)
+            if is_fragile:
```
No, this check only to choose the way of running the tests - from fragile list or not.

avtikhon commented on lib/test.py:
```diff
@@ -250,9 +251,12 @@ def run(self, server):
-            short_status = 'fail'
+                with open(self.tmp_result, mode='rb') as result_file:
+                    result_checksum = md5(result_file.read()).hexdigest()
+            short_status = 'fail:' + result_checksum
```
Added.

avtikhon commented on lib/worker.py:
```diff
@@ -82,13 +83,15 @@ def get_task_groups():
                 'gen_worker': gen_worker,
                 'task_ids': stable_task_ids,
                 'is_parallel': suite.is_parallel(),
+                'is_fragile': False,
```
Seems not needed for now to use additionally enum, code w/o it looks more readable as I think.

avtikhon commented on lib/worker.py:
```diff
@@ -82,13 +83,15 @@ def get_task_groups():
                 'gen_worker': gen_worker,
                 'task_ids': fragile_task_ids,
                 'is_parallel': False,
+                'is_fragile': True,
```
See comment above.

Totktonada replied:
Nice discussion :)

avtikhon commented on dispatcher.py:
```diff
@@ -361,6 +361,7 @@ def __init__(self, key, task_group, randomize):
                 random.shuffle(self.task_ids)
         else:
             self.randomize = False
+        self.is_fragile = task_group['is_fragile']
```
See comment above.

avtikhon commented on lib/test_suite.py:
```diff
@@ -154,22 +176,44 @@ def collect_tests(self):
+                    for checksum in self.ini['fragile']['tests'][test]['checksums']:
+                        checksums.append(checksum)
+                except Exception:
+                    pass
```
Reworked a bit this part, please check if the comment still needed.

avtikhon commented:
Hi, Alexander, please check my comments below.

> 1. The approach with checksums is not universal. It does not work if time-related information appears in a result file. It requires periodical checksums updating if a line number is reported as part of an error. I guess it is okay to operate on a test level and either mark it as stable or as fragile.

Right, some tests that have commands in format like "<failing command> or <print overall information>" and etc. would not pass using this feature, but in all other ways it must help, due to our results files should be the same between runs. It means that this feature is not absolute medicine for flaky tests, but it can help to avoid of it in some places.

> 2. We can either implement reruns within a worker or on the level upward: within the task management system (dispathers-listeners-workers). Since there is nothing tarantool-specific I would look into the latter way. I pushed a variant to `Totktonada/rerun-fragile-test` branch ([0292fe2](https://github.com/tarantool/test-run/commit/0292fe2ad901f33c903387f5498ade4f13acef94)).
> 

It seems too general change, while in my commit for retries it looks much more simple to rerun the test in loop.

> If you want my opinion about extracting checksums from comments — it is disgusting. Sorry.
> 

Ok, sure, I've removed it completely and updated the commits based only on JSON format.

> There are more points I would highlight about the patchset, but it does not have much sense if we'll get rid of checksums at all. The patchset modifies many different parts of test-run: too much for this small feature, IMHO.

In opposite we can't rerun w/o check of results files checksum will hide the new issues.

kyukhin commented:
> I didn't look thoroughly, but I would highlight several points.
> 
> 1. The approach with checksums is not universal. It does not work if time-related information appears in a result file. It requires periodical checksums updating if a line number is reported as part of an error. I guess it is okay to operate on a test level and either mark it as stable or as fragile.

Using time related information in result file is disgusting. However I doubt such things exists to the moment. AFAIU, number of flaky tests supposed to be about 2-3 dozens and we do not update those tests heavily.

> 2. We can either implement reruns within a worker or on the level upward: within the task management system (dispathers-listeners-workers). Since there is nothing tarantool-specific I would look into the latter way. I pushed a variant to `Totktonada/rerun-fragile-test` branch ([0292fe2](https://github.com/tarantool/test-run/commit/0292fe2ad901f33c903387f5498ade4f13acef94)).

I think this is useless over complication. As you justly mentioned, this is simple feature. You approach doesn't look simple at all. I'd stick to Sasha's solution. Moreover your proposal will complicate subsequent changes in part of md5 search and match.
 
> If you want my opinion about extracting checksums from comments — it is disgusting. Sorry.

Agree.

> There are more points I would highlight about the patchset, but it does not have much sense if we'll get rid of checksums at all. The patchset modifies many different parts of test-run: too much for this small feature, IMHO.

I don't agree here. Patchset itself is not that big, it is like 90 lines of code and most changes looks really straightforward.

LeonidVas reviewed (commented):
I agree with A. Turenko that the use of hashsum for fragile tests is a slippery slope. But after some discussion, the Big Boss says:"On monday I'll push it to origin master as is." So, all the proposed changes - up to you. WBR , L. Vasilyev.

Totktonada commented on lib/test_suite.py:
```diff
@@ -110,6 +111,14 @@ def __init__(self, suite_path, args):
+                try:
+                    self.fragile = json.loads(item[1])
+                except ValueError:
+                    print "WARNING: suite '{}' uses old fragile format" . format(suite_path)
```
test-run redirects standard input / output / error streams and so `color_stdout()` is used across the whole codebase. This print will likely excuted before the redirection, but I would stay on the safe side and use `color_stdout()` as usual. BTW, it prints everything line-by-line and so prevent mixing output of different processes in a strange way.

I would accept the old format without deprecation warnings, but it is up to you.

avtikhon replied:
Right, decided to remove.

Totktonada commented on lib/test_suite.py:
```diff
@@ -110,6 +111,14 @@ def __init__(self, suite_path, args):
                 lambda x: os.path.join(suite_path, x),
                 dict.fromkeys(self.ini[i].split())
                 if i in self.ini else dict())
+        for item in config.items("default"):
```
I agree, there is the `self.ini` dictionary.

avtikhon replied:
Both comments implemented.

Totktonada commented on lib/worker.py:
```diff
@@ -327,7 +329,23 @@ def run_loop(self, task_queue, result_queue):
+                # let's run till short_status became 'pass'
+                while short_status != 'pass' and retries_left >= 0:
+                    # print message only after some fails occurred
+                    if re.match(r'^fail:.*', short_status):
```
This commit does not change `short_status` from just `fail` to `fail:<checksum>`, so the regexp looks more as part of the next commit.

avtikhon replied:
Right, corrected.

Totktonada commented on lib/test.py:
```diff
@@ -250,9 +251,12 @@ def run(self, server):
-            short_status = 'fail'
+                with open(self.tmp_result, mode='rb') as result_file:
+                    result_checksum = md5(result_file.read()).hexdigest()
+            short_status = 'fail:' + result_checksum
```
I would pass it in some other way. Say, as a field in `WorkerTaskResult`. Because `short_status` now becomes a structure with a short status and a checksum, but it is not a structure, but a string… it is strange, at least. Not something that a reader of the code would expect.

avtikhon replied:
Corrected as you suggested.

Totktonada commented on lib/test.py:
```diff
@@ -8,6 +8,7 @@
 import sys
 import traceback
 from functools import partial
+from hashlib import md5
```
I don't want to nitpick to much, but the commit header 'Add ability to check failed tests w/ fragile list' does not give the idea what the commit offers. It is like 'everything as before, but with *check*'. 'check' may be almost anything for one, who doesn't read the patch itself.

avtikhon replied:
Corrected.

Totktonada commented:
> Using time related information in result file is disgusting.

It is about places, where an attempt to give more information in case of fail is performed ([example](https://github.com/tarantool/tarantool/blob/f7bcdf4c0405112d390b8ebb6d296717153020ef/test/replication/status.test.lua#L69-L74), consider `or box.info`).

> > We can either implement reruns within a worker or on the level upward: within the task management system (dispathers-listeners-workers).

> I think this is useless over complication.

The queue management is a bit complex part of test-run, yep. It is more appropriate place for this feature, but it is not something critical. I'm okay with the simpler approach.

> > The patchset modifies many different parts of test-run: too much for this small feature, IMHO.
>
> I don't agree here. Patchset itself is not that big, it is like 90 lines of code and most changes looks really straightforward.

I looked a big more into the patchset: it is mostly due to changing of `short_status` in a strange way: it becomes a structure with the status and checksum encoded as a string. So several places need to be updated to extract the actual status from this value.

Totktonada commented:
NB: Don't forget to verify that artifacts from an reexecuted test will be collected.

LeonidVas commented:
> NB: Don't forget to verify that artifacts from an reexecuted test will be collected.

Maybe I don’t understand you, but this is not necessary (with this approach to solving the problem of reruns fragile tests), because re-execution will be run in case of a known failure. If you want to get artifacts for some fragile test - remove the checksum of this test.

Totktonada commented:
> > NB: Don't forget to verify that artifacts from an reexecuted test will be collected.
> 
> Maybe I don’t understand you, but this is not necessary (with this approach to solving the problem of reruns fragile tests), because re-execution will be run in case of a known failure. If you want to get artifacts for some fragile test - remove the checksum of this test.

That's right, but I would not break ability to collect as much statistics as possible if we'll want to do it later.

kyukhin commented:
> > Using time related information in result file is disgusting.
> 
> It is about places, where an attempt to give more information in case of fail is performed ([example](https://github.com/tarantool/tarantool/blob/f7bcdf4c0405112d390b8ebb6d296717153020ef/test/replication/status.test.lua#L69-L74), consider `or box.info`).

So, this things always arise on error path. What's the problem then? We will never see that in .result.

Totktonada commented:
> > > Using time related information in result file is disgusting.
> > 
> > 
> > It is about places, where an attempt to give more information in case of fail is performed ([example](https://github.com/tarantool/tarantool/blob/f7bcdf4c0405112d390b8ebb6d296717153020ef/test/replication/status.test.lua#L69-L74), consider `or box.info`).
> 
> So, this things always arise on error path. What's the problem then? We will never see that in .result.

All those checksums are for output of a failed test.

avtikhon commented on lib/test_suite.py:
```diff
@@ -154,19 +163,27 @@ def collect_tests(self):
+    def get_fragile_list(self):
+        res = []
+        try:
+            res = self.fragile['tests'].keys()
```
Ok, in this way changed the list to dictionary even for the old 'fragile' list.

avtikhon commented on lib/test_suite.py:
```diff
@@ -187,6 +187,15 @@ def fragile_tests(self):
+            for checksum in self.fragile['tests'][test]['checksums']:
+                checksums.append(checksum)
+        except Exception:
+            pass
```
No, some tests without checksums will not be rerun, but its entry will be used to set test in non parallel testing.

avtikhon commented on lib/worker.py:
```diff
@@ -329,7 +330,9 @@ def run_loop(self, task_queue, result_queue, is_fragile):
 
             result_queue.put(self.current_task(task_id))
-            if is_fragile:
+            testname = re.split(r'/', task_id[0])[1]
```
Changed.

avtikhon commented:
> > > > Using time related information in result file is disgusting.
> > > 
> > > 
> > > It is about places, where an attempt to give more information in case of fail is performed ([example](https://github.com/tarantool/tarantool/blob/f7bcdf4c0405112d390b8ebb6d296717153020ef/test/replication/status.test.lua#L69-L74), consider `or box.info`).
> > 
> > 
> > So, this things always arise on error path. What's the problem then? We will never see that in .result.
> 
> All those checksums are for output of a failed test.

Right, but we have only a few of it:
```
grep -RI "or box.info" 2>/dev/null | grep -v ".result:" | grep -v ".reject:" | sed 's#:.*##g' | sort -u | wc -l
9
```

avtikhon commented:
> > NB: Don't forget to verify that artifacts from an reexecuted test will be collected.
> 
> Maybe I don’t understand you, but this is not necessary (with this approach to solving the problem of reruns fragile tests), because re-execution will be run in case of a known failure. If you want to get artifacts for some fragile test - remove the checksum of this test.

Checked.

LeonidVas reviewed (approved):
Looks syntactically correct to me.

Totktonada commented:
I pushed two branches into tarantool today (and rerun failed jobs several times):

- Got several 'address already in use' fails (http_client.test.lua). It is the problem of test-run design (#141), not the test per se.
- Got several fails on net.box_reconnect_after_gh-3164.test.lua with the whole box.info.replication output (time related information, UUIDs).
- Got several fails on net.box_reconnect_after_gh-3164.test.lua with peer UUID, Unix socket path, schema version and so on (will differ from run to run).

There are other fails, but I didn't look whether the output was predictable (at least vinyl.ddl.test.lua, don't remember others).

I know, you're sure you can mark almost everything with checksums, but I see the opposite picture now. Maybe I'm just lucky.

At least, please, left ability to rerun a test, which is not marked with any checksum (I didn't look at the last patchset, maybe it is there already).

Aside of this, the 'transient fail' statistics was nice: no need to roll a screen to find whether reruns occur after full local testing.

avtikhon commented:
> I pushed two branches into tarantool today (and rerun failed jobs several times):
> 
I've run the same 3 days twice in hour and I got really lot of results. Checked some of them I've got the same results as Alexander below, but it was expected. I've updated the table [1] with the new column 'Checksum' and set there some of results, green - checksums that available, red - that have changing data. As I've checked before the same issue with checksum we'll have just in some tests - found only 9 with printing box info and some other tests printing other changing data.

[1] - https://docs.google.com/spreadsheets/d/1LGfHM_6tn84lzlz3XIXQnOqNY_KtzRWuZg33_yuv4YI/edit#gid=1890598391

> * Got several 'address already in use' fails (http_client.test.lua). It is the problem of test-run design (#141), not the test per se.

Right, we have some these fails, but it passes when tests run inline using fragile list for it.

> * Got several fails on net.box_reconnect_after_gh-3164.test.lua with the whole box.info.replication output (time related information, UUIDs).
> * Got several fails on net.box_reconnect_after_gh-3164.test.lua with peer UUID, Unix socket path, schema version and so on (will differ from run to run).
> 

Right this test I've in the table marked as red.

> There are other fails, but I didn't look whether the output was predictable (at least vinyl.ddl.test.lua, don't remember others).
> 
> I know, you're sure you can mark almost everything with checksums, but I see the opposite picture now. Maybe I'm just lucky.
> 

Right, we have some tests like this, but we have 2 ways to resolve it: fix the test either remove box information printing if it is not really needed. I've prepared the initial patch shown as green in the table mentioned above and the results really good - only 1 new fail found [2] ! 

[2] - https://gitlab.com/tarantool/tarantool/-/pipelines/193120097/builds

> At least, please, left ability to rerun a test, which is not marked with any checksum (I didn't look at the last patchset, maybe it is there already).
> 

That was the main cause why this feature with tests reruns I didn't want to have without checksums.

> Aside of this, the 'transient fail' statistics was nice: no need to roll a screen to find whether reruns occur after full local testing.

Actually after checksums will be written to the suites.ini files these fails would be really rare.

Totktonada commented:
> > Got several 'address already in use' fails (http_client.test.lua). It is the problem of test-run design (#141), not the test per se.
>
> Right, we have some these fails, but it passes when tests run inline using fragile list for it.

It is maybe okay as a temporary workaround, but it is not so complex to fix the cause: bind httpd.py on zero port (to let kernel to choose a random one) and read a real port from its output (we already read heartbeats from here).

> > At least, please, left ability to rerun a test, which is not marked with any checksum (I didn't look at the last patchset, maybe it is there already).
>
> That was the main cause why this feature with tests reruns I didn't want to have without checksums.

To be honest, I don't get the answer.

> > Aside of this, the 'transient fail' statistics was nice: no need to roll a screen to find whether reruns occur after full local testing.
>
> Actually after checksums will be written to the suites.ini files these fails would be really rare.

But I still want to know about this, when it occurs, to verify whether it is introduced by a patch I work on. It is okay to postpone this feature, but it would be usable.

Totktonada commented on lib/worker.py:
```diff
@@ -329,24 +332,32 @@ def run_loop(self, task_queue, result_queue, is_fragile):
-            if is_fragile:
+            testname = os.path.basename(task_id[0])
+            fragile_checksums = self.suite.get_test_fragile_checksums(testname)
+            if is_fragile and fragile_checksums:
```
Here you block ability to rerun a test without a checksum. It is possible that a test cannot be verified just using a result file, but we want to allow reruns for it.

I remember the reason why you don't want to enable reruns by default when no checksums are provided: because it means that existing configs will let tests that are currently in a fragile list being retried.

However this is not so until we'll set `retries` explicitly, so it seems we can do this without subtle behaviour changes. I mean, if we'll consider empty (or non-existing) checksums list as 'rerun it anyway', everything will be good. But if you prefer explicit '*' in checksum, I don't mind.

You can also consider this as out of scope (not needed right now), but for me it looks quite natural that…

```json
fragile = {
    "retries": 2,
    "tests": {
        "bitset.test.lua": {
            "issues": ["gh-4095"]
        }
    }
}
```

…means 'rerun bitset.test.lua two times'.

Anyway, it is up to you. I just noted that the problem you highlighted looks as not applicable to our situation.

avtikhon replied:
I've commented it in the same time when you wrote the questions in here:
https://github.com/tarantool/test-run/pull/217#issuecomment-698128881
in short: I agree with it, but as standalone commit for this feature.

Totktonada commented on lib/test_suite.py:
```diff
@@ -265,5 +265,11 @@ def run_test(self, test, server, inspector):
         return self.ini['is_parallel']
 
+    def fragile_retries(self):
+        try:
+            return self.fragile['retries']
+        except Exception:
+            return 0
```
Nit: It is the same as `self.fragile.get('retries', 0)`.

Totktonada replied:
And catching a general exception is considered as the bad practice.

avtikhon replied:
Ok, sure, corrected.

Totktonada commented on lib/test_suite.py:
```diff
@@ -110,6 +111,17 @@ def __init__(self, suite_path, args):
                 lambda x: os.path.join(suite_path, x),
                 dict.fromkeys(self.ini[i].split())
                 if i in self.ini else dict())
+        if config.has_option("default", "fragile"):
+            fragiles = config.get("default", "fragile")
```
Nit: I still don't get why not to use `self.ini.get('fragile', None)`. Since this dictionary is filled, it is used everywhere instead of `config` directly. Mix of `self.ini` and `config` accesses is a bit harder to read, IMHO.

avtikhon replied:
Not possible, because self.ini['fragile'] consists of dictionary instead of string needed for json (because of

self.ini.update(dict(config.items("default")))

below), and convert it to string mess it, check the following change:
```
diff --git a/lib/test_suite.py b/lib/test_suite.py
index d1faa46..cac377a 100644
--- a/lib/test_suite.py
+++ b/lib/test_suite.py
@@ -112,7 +112,8 @@ class TestSuite:
                 dict.fromkeys(self.ini[i].split())
                 if i in self.ini else dict())
         if config.has_option("default", "fragile"):
-            fragiles = config.get("default", "fragile")
+            fragiles = self.ini.get('fragile', None)
+            print("==================================== {}\n" . format(type(fragiles)))
             try:
                 self.fragile = json.loads(fragiles)
                 if 'tests' not in self.fragile:
```
got:
```
==================================== <type 'dict'>

Traceback (most recent call last):
  File "./test-run.py", line 219, in <module>
    status = main_parallel()
  File "./test-run.py", line 134, in main_parallel
    res = main_loop_parallel()
  File "./test-run.py", line 86, in main_loop_parallel
    task_groups = lib.worker.get_task_groups()
  File "/home/avtikhon/Workspaces/tarantool/test-run/lib/worker.py", line 70, in get_task_groups
    suites = find_suites()
  File "/home/avtikhon/Workspaces/tarantool/test-run/lib/worker.py", line 29, in find_suites
    for suite_name in sorted(suite_names)]
  File "/home/avtikhon/Workspaces/tarantool/test-run/lib/test_suite.py", line 118, in __init__
    self.fragile = json.loads(fragiles)
  File "/usr/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
TypeError: expected string or buffer
```

Totktonada replied:
And this dict is strange. Okay.

Totktonada commented on dispatcher.py:
```diff
@@ -388,7 +389,7 @@ def _run_worker(self, worker_id, tcp_port_range):
         color_stdout.queue = self.result_queue
         worker = self.gen_worker(worker_id)
-        worker.run_all(self.task_queue, self.result_queue)
+        worker.run_all(self.task_queue, self.result_queue, self.is_fragile)
```
Several points (not critical, more kind of stylistic thoughts):

1. It seems more natural to pass it to gen_worker() function (because it is property of a worker in fact). If you'll do, don't forget about 'get _id as an arg' comment in worker.py.
2. For worker it is more like 'whether task retries are allowed' property, because it is quite abstract from tests. Maybe we can name it better: say, `enable_retries`.
3. However we can just lean on `retries` suite.ini value, it is available in `run_loop()`. There are two cons. First, we don't need to pass the flag over several calls deep into `run_loop()`. Second, it is quite natural to just follow `retries`.

avtikhon replied:
Right, this flag is not needed and I completely removed it.

Totktonada commented:
Status: It looks almost ready (considering given decisions), I just want to discuss several moments before we'll land the patchset.

avtikhon commented:
> > > Got several 'address already in use' fails (http_client.test.lua). It is the problem of test-run design (#141), not the test per se.
> > 
> > 
> > Right, we have some these fails, but it passes when tests run inline using fragile list for it.
> 
> It is maybe okay as a temporary workaround, but it is not so complex to fix the cause: bind httpd.py on zero port (to let kernel to choose a random one) and read a real port from its output (we already read heartbeats from here).
> 

Absolutely agree that run the tests like this is temporary workaround and it need to be fixed. And we'll continue work on its fixing in already exists standalone issues like:

https://github.com/tarantool/tarantool/issues/4949

> > > At least, please, left ability to rerun a test, which is not marked with any checksum (I didn't look at the last patchset, maybe it is there already).
> > 
> > 
> > That was the main cause why this feature with tests reruns I didn't want to have without checksums.
> 
> To be honest, I don't get the answer.
> 

I mean that blind rerun of the flaky issues is more dangerous than it's manual rerun that we currently have. Check of 'checksums' is a way to make these reruns partly automate to help the testing infrastructure to avoid of manual reruns. Standalone rerun we'll hide the new flaky issues.
Right, we can add the new feature to rerun the needed tests. It can be provided with special checksum/mark like:
   "checksum": [ "temporary_mark_to_rerun_on_any_fail_NOT_TO_COMMIT" ]
and add the special checker in test-run scripts on this mark to pass the tests on rerun on any failed checksum. It will help to check manually some flaky tests. But I think this feature must be committed in separate patch from the current patchset with standalone issue number.
Also in this way I would suggest to have new feature as Sergey Bronnikov suggested to have in some testing mode mandatory reruns for all tests to be able to find the flaky tests at the very start of its appearing, but this feature I would like to implement as standalone commit too.

> > > Aside of this, the 'transient fail' statistics was nice: no need to roll a screen to find whether reruns occur after full local testing.
> > 
> > 
> > Actually after checksums will be written to the suites.ini files these fails would be really rare.
> 
> But I still want to know about this, when it occurs, to verify whether it is introduced by a patch I work on. It is okay to postpone this feature, but it would be usable.

Totktonada reviewed (approved):
LGTM.

Totktonada commented:
The test-run submodule is updated in tarantool in the following commits: [2.6.0-104-g43482eedc](https://github.com/tarantool/tarantool/commit/43482eedc3e139e24e9f12d55761f88ea7f9926c), [2.5.1-86-gc5bb549f6](https://github.com/tarantool/tarantool/commit/c5bb549f6d303c17cb63e4fd6021b22bd09d93b9), [2.4.2-70-gef330c3b0](https://github.com/tarantool/tarantool/commit/ef330c3b06a9fbab7e686347ba94856e614600b7), [1.10.7-35-g91260069c](https://github.com/tarantool/tarantool/commit/91260069c1dc53d86a0c756c2e821e1a9f9e0e95).
